Dynamic switches for tplink

rytilahti opened the pull request:
## This PR is (the first) part of the tplink rewrite series, to be merged into `tplink_rewrite` branch (see #111151).
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
* Create a foundation to use upstream-provided information for creating entities, allowing to expose support for more devices and features while moving some maintenance burden from homeassistant to the upstream library.
* Convert `switch` platform to use the new facilities.
* Expose sub-devices as their own devices that define `via_device`.

TBD:
- [x] TODOs in the code
- [x] Consolidate unique id handling
- [x] Tests
- [x] ~~Documentation update (if it's not better done prior merging the dev branch back to master?)~~

The screenshot below shows two new entities, and one that was previously hardcoded:
![image](https://github.com/home-assistant/core/assets/3705853/67b06237-ceb0-46c9-a69e-c906d58f2b6b)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

rytilahti committed:
Convert SmartDeviceException->KasaException, SmartDevice->Device

rytilahti committed:
Convert SmartDeviceException->KasaException, SmartDevice->Device

rytilahti committed:
Convert SmartDeviceException->KasaException, SmartDevice->Device

rytilahti committed:
Bump version to 0.7.0.dev0

rytilahti committed:
Implement feature-based entities & convert switch platform over

rytilahti committed:
Implement feature-based entities & convert switch platform over

rytilahti committed:
Allow passing kwargs to descriptions

rytilahti committed:
Allow passing kwargs to descriptions

rytilahti committed:
Disable all but explicitly enabled debug features per default

rytilahti committed:
Disable all but explicitly enabled debug features per default

sdb9696 committed:
Migrate tplink integration to new common module kasa library api

sdb9696 committed:
Add has_custom_effects check

sdb9696 committed:
Merge branch 'tplink_rewrite' into tplink_rewrite_merge

sdb9696 committed:
Merge remote-tracking branch 'upstream/tplink_rewrite_switch' into tplink_rewrite_merge

sdb9696 committed:
Remove NO_SWITCH_DEVICE_MODULES const

sdb9696 committed:
Remove led test stuff

sdb9696 committed:
Merge remote-tracking branch 'upstream/tplink_rewrite' into tplink_rewrite_merge

sdb9696 committed:
Remove leftover no switch module const

rytilahti committed:
Cleanup unique_id handling, fix TODOs, and fix typing

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +78,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        key=feature.id,
+        translation_key=feature.id,
+        name=feature.name,
+        # Setting an icon overrides the translation, so we doing it here.
+        # icon=feature.icon,
```
Homeassistant currently picks the `icon` over an icon translation. It would be great to have it the other way around, i.e., to show a relevant icon if there's no translation available, but I couldn't find a way to check for the existence of a translation.

rytilahti replied:
Removed this for now. The proper way is for us to dump a list of available features & add them into the icon translations file.

rytilahti commented on homeassistant/components/tplink/switch.py:
```diff
@@ -32,131 +34,50 @@ async def async_setup_entry(
-        """Turn the child switch on."""
-        await self._plug.turn_on()
+        """Turn the switch on."""
+        await self._feature.set_value(True)  # type: ignore[no-untyped-call]
```
We need to fix this upstream, but it's fine in this development branch to unblock other platforms that depend on the `entity.py` updates done in this PR. Related upstream PR: https://github.com/python-kasa/python-kasa/pull/926

bdraco replied:
It would be better to fix it before merging so we don't have to analyze this code again when we merge back to core

sdb9696 replied:
I think this might not get fixed for this release as the solution for typing the feature values in the library is still up in the air and we might not be able to do it for all use cases. Can we go forwards with this as library limitation that may or may not get resolved?

rytilahti replied:
Alternatively, we could just type with with a union of all values we currently have? `int | float | str | Enum | None` should probably cover it.

bdraco replied:
Seems fine to do an overly broad type for now and than narrow it down later when it could be improved or using a typevar

rytilahti committed:
Add update_before_add and fix a docstring

rytilahti committed:
Call async_update_attrs in __init__

rytilahti commented on homeassistant/components/tplink/switch.py:
```diff
@@ -32,131 +34,51 @@ async def async_setup_entry(
+        self.entity_description = _description_for_feature(
+            SwitchEntityDescription, feature
+        )
         self._async_update_attrs()
```
I thought it'd be fine to avoid calling this in `__init__`, but even adding `update_before_add=True` to `async_add_entities()` does not seem to be enough to get the entity updated for tests, so I suppose this needs to be called.

Or is it better to keep the `_handle_coordinator_update` to call this?

rytilahti replied:
As a side effect for calling the update here makes us not to catch potential exceptions like we are doing in a single call-site for updates, see https://github.com/python-kasa/python-kasa/pull/933

bdraco replied:
> I thought it'd be fine to avoid calling this in `__init__`, but even adding `update_before_add=True` to `async_add_entities()` does not seem to be enough to get the entity updated for tests, so I suppose this needs to be called.
> 
> 
> 
> Or is it better to keep the `_handle_coordinator_update` to call this?

Update before add is highly undesirable because it slows down platform setup and makes everything much more complicated since it has to do another update after the first refresh

rytilahti replied:
Thanks for the explanation, I'll look into what's the best way to handle this nicely.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +78,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        self._feature = feature
         self._attr_device_info = DeviceInfo(
-            connections={(dr.CONNECTION_NETWORK_MAC, device.mac)},
+            # connections={(dr.CONNECTION_NETWORK_MAC, device.mac)},
```
Open question if this is needed when we have identifiers? If set for child devices, all of them will be presented by a single device.

bdraco replied:
If its a child device we shouldn't set it and instead use `via_device`

rytilahti replied:
Done. `via_device` is now used when we have a parent device, otherwise the connection is created.

rytilahti committed:
Define connections when the device has no parent

rytilahti commented:
_By the power of circuits and code,
Rise, review bot, from your digital abode.
Analyze and critique with sharp insight,
Bring your wisdom into our sight._
- chatgpt

@coderabbitai review

rytilahti committed:
Convert match-case to map, remove setting the icon for now

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,204 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+                _LOGGER.error(
+                    "Unhandled category %s, fallback to DIAGNOSTIC", feature.category
+                )
+                return EntityCategory.DIAGNOSTIC
```
Done.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+
+    def _category_for_feature(self, feature: Feature) -> EntityCategory | None:
+        """Return entity category for a feature."""
+        category_map = {
```
Thanks for the review, done :+1:

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+            return False
+
+        # We skip primary features for device types that have specialized platforms.
+        ignore_primary_controls_devicetypes = [
```
Done.

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -2,20 +2,42 @@
+    entity_registry_enabled_default: bool
+
+
+_LOGGER = logging.getLogger(__name__)
```
Usually these go right after imports

rytilahti replied:
Done.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+
+    This is a helper that calls *_entities_for_device* for the device and its children.
+    """
+    entities = []
```
Added.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        """Handle updated data from the coordinator."""
+        try:
+            self._async_update_attrs()
+            self._attr_available = True
```
Done.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+    return [
+        entity_class(device, coordinator, feat, parent=parent)
+        for feat in device.features.values()
+        if _filter(device, feat)
```
Yeah, that'd be faster but harder to read. I usually like to avoid too complex checks on comprehensions as it not only affects readability, but also makes it harder to set breakpoints for debugging.

I'm open for options to make this nicer :-)

bdraco replied:
```python
    return [
        entity_class(device, coordinator, feat, parent=parent)
        for feat in device.features.values()
        if feat.type == feature_type
        and (
            feat.category != Feature.Category.Primary
            or device.device_type not in DEVICETYPES_WITH_SPECIALIZED_PLATFORMS
        )
    ]
```

Assuming this is what is intended

rytilahti replied:
Done, although I don't think it's worth for readability reasons.

rytilahti committed:
Address code review comments

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +95,189 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        # The rest of the initialization takes care of setting a proper unique_id
+        # This is transitional and will become cleaner as future platforms get converted.
+
+        # If the unique id is already defined, we don't need to do anything.
+        if self._attr_unique_id is not None:
+            return
+
```
Its a bit unexpected that the check is based on unique id and not something more deterministic

rytilahti replied:
Yeah, I'll find a better solution for that, likely just by adding a boolean flag that is set to true by the specialized platforms.

rytilahti replied:
Moved all unique_id logic into `_get_unique_id` of the base entity class to have it in a single place.

rytilahti committed:
Merge branch 'tplink_rewrite' into tplink_rewrite_switch

rytilahti committed:
Fix handling of primary controls

sdb9696 commented on homeassistant/components/tplink/switch.py:
```diff
@@ -32,131 +34,51 @@ async def async_setup_entry(
-    """Representation of a TPLink Smart Plug switch."""
-
-    _attr_name: str | None = None
+class Switch(CoordinatedTPLinkEntity, SwitchEntity):
```
```suggestion
class TPLinkSwitch(CoordinatedTPLinkEntity, SwitchEntity):
```
I think we agreed to stick with naming the platforms `TPLinkBlah` the other day.

sdb9696 commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +94,193 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        feature: Feature | None = None,
+        parent: Device | None = None,
```
I'm wondering if these should be `kwargs` only?

rytilahti replied:
Changed.

rytilahti committed:
Add icon translations for auto_update_enabled and auto_off_enabled

rytilahti committed:
Rename switch to tplinkswitch

rytilahti committed:
Pass only device & coordinator as position, rest are kwargs only

rytilahti committed:
Use consistent ordering for positional arguments (device, coordinator)

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +94,194 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
-        """Initialize the switch."""
+        """Initialize the entity."""
         super().__init__(coordinator)
         self.device: Device = device
```
Perhaps we should rename this to `self._device` for consistency?

rytilahti replied:
Renamed.

rytilahti committed:
Pass parent for sensor entities to avoid creating invalid connections

This is necessary as CoordinatedTPLinkEntity checks for parent to decide whether to use identifiers or connections.

rytilahti committed:
Merge branch 'tplink_rewrite' into tplink_rewrite_switch

rytilahti committed:
Merge branch 'tplink_rewrite' into tplink_rewrite_switch

rytilahti commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -111,11 +111,11 @@ def async_emeter_from_device(
 def _async_sensors_for_device(
```
Changes into this platform were necessary in this PR to avoid setting up a connection on child device sensors, as [we check for `parent` to decide if to use a connection or an identifier for the entity](https://github.com/home-assistant/core/pull/117758/files#diff-3039452aefa70a1099b307cc1eea25b60996f064fa974f679a56c732e09919a8R123).

rytilahti committed:
Handle some review comments

* Inline entity filter
* Allow entity classes to pass legacy unique_id handling as parameter

sdb9696 committed:
Add primary child entities to parent device (#119200)

Adds primary entities to the parent device for devices such as power strips and wall switches.

rytilahti committed:
Merge branch 'tplink_rewrite' into tplink_rewrite_switch

rytilahti committed:
Rename device to _device

rytilahti committed:
Consolidate unique_id creation inside a single function for easier refactoring later on

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +95,213 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+                # alias makes more sense in the context of the parent.
+                # i.e. Hall Ceiling Fan & Bedroom Ceiling Fan; Child device aliases will be Ceiling Fan
+                # and Dimmer Switch for both so should be distinguished by the parent name.
+                name = f"{parent.alias} {device.alias}"
```
missing coverage

sdb9696 replied:
Fixed

sdb9696 replied:
Fixed

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +95,213 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        if isinstance(self, LightEntity):
+            unique_id = device.mac.replace(":", "").upper()
+            # For backwards compat with pyHS100
+            if device.device_type == DeviceType.Dimmer:
```
If these are always Enums and not StrEnum we can use `is` instead of `==`

sdb9696 replied:
Fixed

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +95,213 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        if (
+            entity_category := FEATURE_CATEGORY_TO_ENTITY_CATEGORY.get(feature.category)
+        ) is None:
+            _LOGGER.error(
```
missing coverage

sdb9696 replied:
Fixed

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -57,24 +95,213 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+            self._async_update_attrs()
+        except Exception as ex:  # noqa: BLE001
+            if self._attr_available:
+                _LOGGER.warning(
```
missing coverage

sdb9696 replied:
Fixed

sdb9696 committed:
Fix tplink attribute update handling, test coverage and other review comments (#119374)

rytilahti committed:
Make the example switches translatable

sdb9696 reviewed (approved):
LGTM!

bdraco commented on tests/components/tplink/test_init.py:
```diff
@@ -297,3 +298,101 @@ async def test_plug_auth_fails(hass: HomeAssistant) -> None:
+    config_entry.add_to_hass(hass)
+    light = _mocked_device(modules=[Module.Light], alias="my_light")
+    light_module = light.modules[Module.Light]
+    p = PropertyMock(side_effect=KasaException)
+    type(light_module).color_temp = p
```
Please avoid using single char variable names outside of a simple genexpr or comprehension

bdraco commented on tests/components/tplink/test_init.py:
```diff
@@ -297,3 +298,101 @@ async def test_plug_auth_fails(hass: HomeAssistant) -> None:
+    state = hass.states.get(entity_id)
+    assert state.state == STATE_ON
+
+    p = PropertyMock(side_effect=KasaException)
+    type(light_module).color_temp = p
```
Please avoid using single char variable names outside of a simple genexpr or comprehension
